### ISSUE TITLE
do notation in right operand looks ambiguous

### ISSUE THREAD

**Issue by roberth** [tweag/ormolu#528]
Ormolu formats to this code, which isn't very clear:

```
y = do
  _ <- Nothing
  undefined
  <|> do
    Just 2
```

If you're not intimately familiar with `do` syntax, you may wonder whether `y` is `Nothing` or `Just 2`.

This layout is easier to understand:

```
y = do
    _ <- Nothing
    undefined
  <|> do
    Just 2
```

**Comment by mrkkrp** [comment]
I think we should try to shift the infix operator one indentation step to the left in this particular case. As the last resort we could also shift half-step, but we need a good reason to do that.

**Comment by tbagrel1** [comment]
@roberth, yes, I suppose the diff is smaller when going from single to multi-line `do` if we stay consistent for both single-line and multi-line `do` blocks.

I currenlty have a hard time figuring out what would be the desired output of a chain of {`do` block} `op` {`do` block} `op` {`do` block}:

Option 1:
```haskell
x = do
  something
 <> do
  something
 <> do
  something
```

Option 2:
```haskell
x = do
  something
 <> do
    something
 <> do
    something
```

Option 3:
```haskell
x = do
  something
 <> do
    something
   <> do
      something
```

**Comment by roberth** [comment]
Shouldn't the first `<>` in your current implementation be non-indented because it is technically in a hanging context?

```haskell
x =
  do
    something
   <> do    -- unnecessary indent
      something
    <> do   -- correct given context but still indented too much because of first <>'s rhs
      something
    <> do   -- correct given context but still indented too much because of first <>'s rhs
      something
```

**Comment by mrkkrp** [comment]
Good question. It's hard to tell :-D Seriously though, It looks like `(:)` will be applied to the whole do block. And yes, that's what I had in mind. Simply because it is easier to spot `(:)` than deduce that all operands are do blocks (what to do with a mixed sequence?).

**Comment by tbagrel1** [comment]
At the beginning, I tried something like this, which would work even with mixed sequences
```haskell
x =
  1 :
  do
    2
   :
  do
    3
   :
  do
    [4]
```
But I got parsing errors when the do block is in hanging position, and it would be hard to implement cleanly

**Comment by mrkkrp** [comment]
Can we have this rule: if at least one operand in an operator chain is a do block, then do this:

```haskell
x =
  1
 : do
  blah
  blah
 : do
  foo
  bar
```

Frankly, I think this is not a common kind of expression, so we can do whatever is easier to implement in a consistent way.
